### ISSUE TITLE
Nano: Refactor inference API from pl model to trainer

### ISSUE THREAD

**Issue by zhentaocc** [intel-analytics/BigDL#4265]
# What the refactoring for 
1. We have an `pl_module.inference(..)` entry to inference by a list of tensors or numpy arrays as below.
https://github.com/intel-analytics/BigDL/blob/main/python/nano/src/bigdl/nano/pytorch/runtime_binding/base_inference.py#L62-L155   
To add openvino support, I have to add another if sentence for this new backend even though the inference loop is already implemented in other backends(onnx, pytorch). In case we will have other backend engines to be integrated into nano, I think we should decouple this implementation. into 3 or more parts:
- `on_inference_start()`
- `inference_loop()`
- `on_inference_end()` 

This is similar to what pytorch lightning provides (train, test, predict, validate). When a new backend is to be added, we just need to modify `on_inference_start()` and `on_inference_end()` according to your feature request.   

2. This is also motivated to comply with Pytorch-Lighting usage style. We have `trainer.fit()`, `trainer.train()`, `trainer.test()`, `trainer.validate()` to take over all forward loop related works,  while `pl_module.inference()` seems more "Keras-Style". So maybe `trainer.inference()` instead will be more straightforward for pytorch or pytorch-lightning users.  
On the other hand, by putting inference in trainer, we can take advantage of PL callback feature as stated above with `on_xxx_start` and `on_xxx_end`.
# What to do 
1.  `inference` can be moved from pl_model to trainer. With current usage, no matter if your model is a pytorch or PL model, you must call `trainer.compile()` to bind `inference` to the model and use it as below.
```python
trainer = Trainer()
trainer.compile(model, openvino=True)
model.inference(data, ...)
```
If we move it to trainer, it can be:
```python
trainer = Trainer(openvino=True)
trainer.inference(model, data, ...)
```
2. Refactor  https://github.com/intel-analytics/BigDL/blob/main/python/nano/src/bigdl/nano/pytorch/runtime_binding/base_inference.py into callback style. Make sure it can accommodate new backends easily.
3. Refactor https://github.com/intel-analytics/BigDL/tree/main/python/nano/src/bigdl/nano/pytorch/runtime_binding 
4. After `trainer.inference(..)` is ready, then we can remove `pl_module.inference(..)`

**Comment by zhentaocc** [comment]
@jason-dai any thoughts? I will firstly use this method on openvino and won't make any changes to what we have on inference now.

**Comment by zhentaocc** [comment]
An example for what I did for openvino: https://github.com/intel-analytics/BigDL/pull/4256/files

**Comment by TheaperDeng** [comment]
![image](https://user-images.githubusercontent.com/35031544/159211760-c704a184-c1ee-45f7-86de-1c79f82049a2.png)
![image](https://user-images.githubusercontent.com/35031544/159211725-35ca05e0-30a7-4160-a95c-954b9bb411fd.png)

**Comment by jason-dai** [comment]
I think we should make sure the APIs for OpenVINO and ONNX RT consistent with each other (including `model.eval_onnx`, `model.eval_openvino`, `trainer.compile`, `inference`), instead of introducing new APIs such as `trainer = Trainer(openvino=True)`
